src/v/datalake/record_schema_resolver.cc

```diff
@@ -330,4 +334,136 @@ record_schema_resolver::resolve_identifier(schema_identifier ident) const {
       from_identifier_visitor{std::move(ident), shared_schema});
 }
 
+latest_protobuf_schema_resolver::latest_protobuf_schema_resolver(
+  schema::registry& sr,
+  model::topic_view topic_name,
+  ss::sstring full_message_name,
+  ss::lowres_clock::duration cache_duration,
+  std::optional<std::reference_wrapper<schema_cache>> sc)
+  : sr_(&sr)
+  , subject_(fmt::format("{}-value", topic_name()))
+  , full_message_name_(std::move(full_message_name))
+  , cache_duration_(cache_duration)
+  , cache_(sc) {}
+
+namespace {
+
+struct pb_descriptor_lookup_result {
+    std::reference_wrapper<const google::protobuf::Descriptor> descriptor;
+    std::vector<int32_t> offsets;
+};
+
+checked<pb_descriptor_lookup_result, type_resolver::errc> lookup_descriptor(
+  const ppsr::protobuf_schema_definition& pb_def,
+  std::string_view message_full_name) {
+    auto d_res = ppsr::descriptor(pb_def, message_full_name);
+    if (d_res.has_error()) {
+        return type_resolver::errc::invalid_config;
+    }
+    // Build up the offsets by walking the descriptor tree
+    std::vector<int> offsets;
+    for (const google::protobuf::Descriptor* d = &d_res.value().get();
+         d != nullptr;
+         d = d->containing_type()) {
+        offsets.push_back(d->index());
+    }
+    std::ranges::reverse(offsets);
+    return pb_descriptor_lookup_result{
+      .descriptor = d_res.value(),
+      .offsets = std::move(offsets),
+    };
+}
+
+} // namespace
+
+ss::future<checked<type_and_buf, type_resolver::errc>>
+latest_protobuf_schema_resolver::resolve_buf_type(
+  std::optional<iobuf> b) const {
+    // Prevent multiple fibers from resolving the protobuf. Since the subject
+    // is fixed there would be no harm in letting them race either, but it seems
+    // better to not waste work.
```

Comment:
FWIW each resolver on the data path is only used by a single translator, so there _shouldn't_ be these kinds of races unless we begin to share the cache.

There is a shared schema cache, but it's the S3 fifo cache, not sure it lends itself well to having a TTL for schemas

Reply:
Thanks, removed